libpod/runtime_pod_linux.go

```diff
@@ -214,34 +214,37 @@ func (r *Runtime) removePod(ctx context.Context, p *Pod, removeCtrs, force bool,
 		return fmt.Errorf("pod %s contains containers and cannot be removed: %w", p.ID(), define.ErrCtrExists)
 	}
 
-	// Go through and lock all containers so we can operate on them all at
-	// once.
-	// First loop also checks that we are ready to go ahead and remove.
-	containersLocked := true
+	ctrNamedVolumes := make(map[string]*ContainerNamedVolume)
+
+	var removalErr error
 	for _, ctr := range ctrs {
-		ctrLock := ctr.lock
-		ctrLock.Lock()
-		defer func() {
-			if containersLocked {
+		err := func() error {
+			ctrLock := ctr.lock
+			ctrLock.Lock()
+			defer func() {
 				ctrLock.Unlock()
+			}()
+
+			if err := ctr.syncContainer(); err != nil {
+				return err
 			}
-		}()
 
-		// If we're force-removing, no need to check status.
-		if force {
-			continue
-		}
+			for _, vol := range ctr.config.NamedVolumes {
+				ctrNamedVolumes[vol.Name] = vol
+			}
 
-		// Sync all containers
-		if err := ctr.syncContainer(); err != nil {
-			return err
-		}
+			return r.removeContainer(ctx, ctr, force, false, true, timeout)
+		}()
 
-		// Ensure state appropriate for removal
-		if err := ctr.checkReadyForRemoval(); err != nil {
-			return fmt.Errorf("pod %s has containers that are not ready to be removed: %w", p.ID(), err)
+		if removalErr == nil {
+			removalErr = err
+		} else {
+			logrus.Errorf("Removing container %s from pod %s: %v", ctr.ID(), p.ID(), err)
 		}
 	}
+	if removalErr != nil {
+		return removalErr
+	}
 
 	// We're going to be removing containers.
 	// If we are Cgroupfs cgroup driver, to avoid races, we need to hit
```

Comment:
Should this block be moved up, to ensure we're still preventing cleanup processes?

Reply:
I think that would require to lock all containers at the same time which would reintroduce the deadlock.

@giuseppe WDYT?

Reply:
could we just remove this hack?

Since there is no locking now for removing containers, the cleanup processes should not hang for too long so it is fine if they run.

I've pushed a new version without this code block; let's see if it causes any issue in the CI.